test/src/history/code_history_controller_test.dart

```diff
@@ -0,0 +1,369 @@
+        await wt.moveCursor(-1); // Creates.
+
+        controller.value = controller.value.typed('c');
+
+        await wt.sendUndo(); // Creates.
+
+        expect(controller.historyController.stack.length, 4);
+        expect(controller.fullText, MethodSnippet.full + 'ba');
+        expect(
+          controller.selection,
+          const TextSelection.collapsed(
+            offset: MethodSnippet.visible.length + 1,
+          ),
+        );
+
+        await wt.sendUndo();
+
+        expect(controller.fullText, MethodSnippet.full + 'a');
+
+        await wt.sendUndo(); // To initial.
+
+        expect(controller.fullText, MethodSnippet.full);
+        expect(
+          controller.selection,
+          const TextSelection.collapsed(offset: -1),
+        );
+
+        await wt.sendUndo(); // No effect.
+
+        expect(controller.fullText, MethodSnippet.full);
+
+        await wt.sendRedo();
+
+        expect(controller.fullText, MethodSnippet.full + 'a');
+        expect(
+          controller.selection,
+          const TextSelection.collapsed(
+            offset: MethodSnippet.visible.length + 1,
+          ),
+        );
+
+        await wt.sendRedo();
+
+        expect(controller.fullText, MethodSnippet.full + 'ba');
+
+        await wt.sendRedo();
+
+        expect(controller.fullText, MethodSnippet.full + 'cba');
+
+        await wt.sendRedo(); // No effect.
+
+        expect(controller.fullText, MethodSnippet.full + 'cba');
+      });
+
+      testWidgets('Changing disables redo', (WidgetTester wt) async {
+        final controller = await pumpController(wt, MethodSnippet.full);
+        await wt.cursorEnd();
+
+        controller.value = controller.value.typed('a');
+        await wt.moveCursor(-1); // Creates.
+
+        controller.value = controller.value.typed('b');
+
+        await wt.sendUndo(); // Creates.
+
+        expect(controller.fullText, MethodSnippet.full + 'a');
+        expect(
+          controller.selection,
+          const TextSelection.collapsed(
+            offset: MethodSnippet.visible.length + 1,
+          ),
+        );
+
+        controller.value = controller.value.typed('b'); // Deletes redo records.
+
+        expect(controller.historyController.stack.length, 2);
+
+        await wt.sendRedo(); // No effect.
+
+        expect(controller.historyController.stack.length, 2);
+
+        await wt.moveCursor(-1); // Creates.
+
+        expect(controller.fullText, MethodSnippet.full + 'ab');
+        expect(
+          controller.selection,
+          const TextSelection.collapsed(
+            offset: MethodSnippet.visible.length + 1,
+          ),
+        );
+
+        await wt.sendUndo();
+        await wt.sendUndo();
+
+        expect(controller.fullText, MethodSnippet.full);
+
+        await wt.sendRedo();
+        await wt.sendRedo();
+
+        expect(controller.fullText, MethodSnippet.full + 'ab');
+      });
+
+      testWidgets('Limit depth', (WidgetTester wt) async {
```

Comment:
It will be better to split on three methods: `limitFillsAsExpected`, `undoLimitAsExpected`, `redoAsExpected`

Reply:
Outside of tests this would sure need to be splitted.

But with tests it is more important to see and run each one as a whole. So splitting is less aggressive in tests. This is because grasping `expect` outside of `test`/`testWidgets` makes it hard to follow.

So instead of extracting the sections I added comments between them.